pkg/peer/subscription/rtp.go

```diff
@@ -0,0 +1,135 @@
+type PacketRewriter struct {
+	// The SSRC of the previously forwarded packet.
+	previouslyForwardedSSRC uint32
+	// SSRC that we're using for all packets that we're forwarding.
+	// This is the SSRC that we're sending to the remote peer. Typically,
+	// this is the SSRC of the lowest layer for the simulcast track.
+	outgoingSSRC uint32
+	// The highest identifiers of the forwarded packet,i.e. the IDs of
+	// the **latest** (in terms of timestamp and number) packet. This is not
+	// necessarily the IDs of the **last forwarded packet**, due to packets
+	// that may arrive out-of-order.
+	maxOutgoingIDs PacketIdentifiers
+	// The identifiers of the **first incoming packet** after swithing
+	// layers. We use it to calulcate their relative position in the RTP stream.
+	firstIncomingIDs PacketIdentifiers
+	// The identifiers of the **first outgoing packet** after switching
+	// layers. This is our "base" to calculate the proper timestamp when
+	// forwarding/rewriting the packet.
+	firstOutgoingIDs PacketIdentifiers
+}
+
+// Creates a new instance of the `PacketRewriter`.
+func NewPacketRewriter(outgoingSSRC uint32) *PacketRewriter {
+	rewriter := new(PacketRewriter)
+	rewriter.outgoingSSRC = outgoingSSRC
+	return rewriter
+}
+
+// Process new incoming packet.
+func (p *PacketRewriter) ProcessIncoming(packet *rtp.Packet) (RewrittenRTPPacket, error) {
+	// Store current incoming packet identifiers.
+	incomingIDs := PacketIdentifiers{packet.Timestamp, packet.SequenceNumber}
+
+	// Calculated outgoing IDs of the current packet.
+	outgoingIDs := PacketIdentifiers{0, 0}
+
+	// Check if we've just switched the layer before this packet, i.e. if
+	// it is the first packet after switching layers.
+	if p.previouslyForwardedSSRC != packet.SSRC {
+		// There is a gap of 1 seqnum to signify to the decoder that the previous frame
+		// was (probably) incomplete. That's why there's a 2 for the seqnum.
+		delta := PacketIdentifiers{1, 2}
+
+		// We make an exception for the very first packet that we're forwarding
+		// so that we start with 0 seqnum and 0 timestamp for the first packet.
+		if p.previouslyForwardedSSRC == 0 {
+			delta = PacketIdentifiers{0, 0}
+		}
+
+		// Calculate the outgoing IDs of the current packet
+		// as well as our new "base" for calculation of timestamps.
+		outgoingIDs = p.maxOutgoingIDs.Add(delta)
+		p.firstIncomingIDs = incomingIDs
+		p.firstOutgoingIDs = outgoingIDs
+
+		// Update the SSRC of the previously forwarded packet.
+		p.previouslyForwardedSSRC = packet.SSRC
+	} else {
+		// If the incoming timeestamp or sequence number are smaller than the timestamp of the first packet after the switch,
+		// then we're getting the old packet (before switch), which we're not interested in, so we drop it.
+		if incomingIDs.timestamp < p.firstIncomingIDs.timestamp ||
+			incomingIDs.sequenceNumber < p.firstIncomingIDs.sequenceNumber {
+			return nil, fmt.Errorf("Ignoring old packet")
+		}
+
+		delta := incomingIDs.Sub(p.firstIncomingIDs)
```

Comment:
Bear in mind that these IDs will wrap around, SSRCs quite frequently, so this will end up being negative sooner or later.

Reply:
The timestamps and sequence numbers? Won't they be caught by the `if` above?

Sorry, I did not quite get what you mean by wrapping around SSRCs though.

Reply:
Oh sorry, I meant seqnums not SSRCs.

Reply:
Do you by chance know what's the right way to rewrite these values once they wrap? I.e. what do actually browsers expect? I'm not that much concerned about the timestamps since they are `uint32` and so far I've never experienced them wrapping (though I believe it might happen for longer conferences?), but the wrapping of sequence numbers does happen for conferences e.g. conferences that last for half an hour.

Given that the sequence number is just an `uint16`, we reach this rather sooner (2^16=65536). Do you by chance know if the browser would expect a sequence number of `0` after we've sent the one with the sequence number `65535`?

Reply:
Yeah, it just wraps around in the normal way, so 0 comes after 65535. Bear in mind that timestamps don't start from 0 though, so it's not just length of call that's the factor, it's also being unlucky and it choosing a starting TS near the top of the range.